### PR TITLE
DAOS-13707 doc: add libpil4dfs in filesystem.md

### PR DESCRIPTION
**Pull request by wiliamhuang**
Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: docs/user/filesystem.md [hunk 1]
```diff
@@ -894,3 +894,126 @@ $
 
 While this mode is not expected to be used directly by users, it is useful for
 the unified namespace integration.
+
+## Interception Library `libpil4dfs`
+
+A new interception library called `libpil4dfs` is introduced. It is similar to libioil,
```
Comment (mchaarawi) [inline docs/user/filesystem.md:900]: please don't use new, introduced, etc. just speak about libpil4dfs. these are not release notes and should not talk about introducing something in a particular release as this lives with the source code.

File: docs/user/filesystem.md [hunk 2]
```diff
@@ -894,3 +894,126 @@ $
...
+but it intercepts more APIs, including read/write as well as metadata related functions.
+This provides kernel-bypass for I/O data, leading to improved performance. This is a preview
```
Comment (mchaarawi) [inline docs/user/filesystem.md:902]: this sounds copy paste from the ioil section ..

File: docs/user/filesystem.md [hunk 3]
```diff
@@ -894,3 +894,126 @@ $
...
+   Operation                     Max            Min           Mean        Std Dev
+   ---------                     ---            ---           ----        -------
+   File creation                2442.835       2442.835       2442.835          0.000
+   File stat                       0.000          0.000          0.000          0.000
+   File read                       0.000          0.000          0.000          0.000
+   File removal                    0.000          0.000          0.000          0.000
+   Tree creation                2957.901       2957.901       2957.901          0.000
+   Tree removal                    0.000          0.000          0.000          0.000
+-- finished at 06/20/2023 18:57:55 --
+
```
Comment (mchaarawi) [inline docs/user/filesystem.md:952]: no need to include all this mdtest output.

File: docs/user/filesystem.md [hunk 4]
```diff
@@ -894,3 +894,126 @@ $
...
+[seek   ]  1667
+[mkdir  ]  2
+[rmdir  ]  0
+[rename ]  0
+[mmap   ]  0
+
+[op_sum ]  5003
+```
+
+### Using libpil4dfs with env
```
Comment (mchaarawi) [inline docs/user/filesystem.md:971]: please remove this section, or just mention this is for testing purposes only and not well tested/supported.

File: docs/user/filesystem.md [hunk 5]
```diff
@@ -894,3 +894,126 @@ $
 
 While this mode is not expected to be used directly by users, it is useful for
 the unified namespace integration.
+
+## Interception Library `libpil4dfs`
+
+A new interception library called `libpil4dfs` is introduced. It is similar to libioil,
+but it intercepts more APIs, including read/write as well as metadata related functions.
+This provides kernel-bypass for I/O data, leading to improved performance. This is a preview
+version. Some features are not implemented yet.
```
Comment (mchaarawi) [inline docs/user/filesystem.md:903]: please move the last sentence to the limitation section which can be removed / updated once this is not a preview anymore.

**Review by mchaarawi — CHANGES_REQUESTED** [review]

File: docs/user/filesystem.md (hunk 1 above)
Comment (wiliamhuang) [reply docs/user/filesystem.md]: Revised. Thank you!

File: docs/user/filesystem.md (hunk 2 above)
Comment (wiliamhuang) [reply docs/user/filesystem.md]: Removed the copied sentences. Thank you!

File: docs/user/filesystem.md (hunk 3 above)
Comment (wiliamhuang) [reply docs/user/filesystem.md]: mdtest output is removed.

File: docs/user/filesystem.md (hunk 4 above)
Comment (wiliamhuang) [reply docs/user/filesystem.md]: This section is removed.

File: docs/user/filesystem.md (hunk 5 above)
Comment (wiliamhuang) [reply docs/user/filesystem.md]: This sentence is moved to limitation section as suggested. Thank you!

**Review by Michael-Hennecke — APPROVED** [review]

**Pull request merged**